src/core/net/dns_client.cpp

```diff
@@ -1106,6 +1191,121 @@ void Client::HandleTimer(void)
     }
 }
 
+#if OPENTHREAD_CONFIG_DNS_OVER_TCP_ENABLE
+void Client::HandleTcpTimer(void)
+{
+    // If message wasn't received in timeout time close connection.
+    mEndpoint.SendEndOfStream();
+    mTcpInitialized = false;
+}
+
+void Client::HandleTcpEstablished(otTcpEndpoint *aEndpoint)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    mEndpoint.SendByReference(mSendLink, 0);
+}
+
+void Client::HandleTcpEstablishedCallback(otTcpEndpoint *aEndpoint)
+{
+    LogDebg("Dns tcp connection established.\n");
+    static_cast<Client *>(otTcpEndpointGetContext(aEndpoint))->HandleTcpEstablished(aEndpoint);
+}
+
+void Client::HandleTcpSendDoneCallback(otTcpEndpoint *aEndpoint, otLinkedBuffer *aData)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    OT_UNUSED_VARIABLE(aData);
+}
+
+void Client::HandleTcpReceiveAvailable(otTcpEndpoint *aEndpoint,
+                                       size_t         aBytesAvailable,
+                                       bool           aEndOfStream,
+                                       size_t         aBytesRemaining)
+{
+    OT_UNUSED_VARIABLE(aEndpoint);
+    OT_UNUSED_VARIABLE(aBytesRemaining);
+
+    const otLinkedBuffer *data;
+    size_t                totalReceived = 0;
+    Message              *message       = nullptr;
+    Response              response;
+    QueryType             type;
+    Error                 responseError;
+    uint16_t              length = 0;
+
+    if (aEndOfStream)
+    {
+        // Cleanup is done in dissconnected callback.
+        mTcpTimer.Stop();
+    }
+
+    VerifyOrExit(aBytesAvailable != 0);
+    mEndpoint.ReceiveByReference(data);
+    message = mSocket.NewMessage(0);
+    VerifyOrExit(message != nullptr);
+
+    if (data != nullptr)
```

Comment:
The `otLinkedBuffer` we get from `ReceiveByReference()` can contain a chain of buffers and not just one. We should check and process each one (similar how CLI example does)
https://github.com/openthread/openthread/blob/117c8519eceb430ba0bec503c6b0fc17418dfa8f/src/cli/cli_tcp.cpp#L520-L526

Reply:
Are You sure about this? I have allocated just one buffer for the reception.
Does TCP stack allocates something extra?

Reply:
@abtink I have applied your suggestions please check if it's ok.

Reply:
I think we probably need more cases here, like if we have more bytes than what we expect to read (in case we get multiple DNS responses in one callback).